### ISSUE TITLE
[Bug] Crash Firestore Unity Android

### ISSUE THREAD

**Issue by pelevinma** [firebase/quickstart-unity#1080]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2020.3.12
  * Firebase Unity SDK version: 8.0.0
  * Source you installed the SDK: .unitypackage
  * Problematic Firebase Component: Firestore
  * Other Firebase Components in use: Analytics, Auth, Crashlytics
  * Additional SDKs you are using: Facebook, AdMob
  * Platform you are using the Unity editor on: Mac
  * Platform you are targeting: Android
  * Scripting Runtime: IL2CPP

### [REQUIRED] Please describe the issue here:

1) Updated Firebase from 7.2.0 to 8.0.0
2) Run Android Resolcer -> Force Resolve
2) Compile and start Android build
3) Trying to save progress to Firestore server
4) I get a 100% crash:
_Java class com/google/firebase/firestore/internal/cpp/QueryEventListener not found.  Please verify the AAR which contains the com/google/firebase/firestore/internal/cpp/QueryEventListener class is included in your app._

Could not run "Unity quickstarts" project on Unity 2020.3, got many errors when opening the project

#### Relevant Code:

DocumentReference docRef = db.Collection("users_v2").Document(userId);
Dictionary<string, object> userData = new Dictionary<string, object>
{
        // my user data
};

// crash in next line
docRef.SetAsync(userData).ContinueWithOnMainThread(task =>
{
    // callback
});

**Comment by dconeybe** [comment]
Hi @pelevinma. I suspect something has gone wrong with the upgrade. Do you still see this issue if you just install 8.0.0 initially (and bypass the upgrade path)?

**Comment by pelevinma** [comment]
Yes, still relevant.
I tried:
1) completely uninstall old firebase and install from scratch
2) update current firebase from 7.2.0 to 8.0.0
- removing 7.2.0 files via Google version Handler
- without removing 7.2.0 files

To check this on a clean project, I need to attach the google-services.json file and use the bundleId from the main project, right?

**Comment by dconeybe** [comment]
Yes, that sounds correct. I'll look into this further and reply in the next day or two.

**Comment by pelevinma** [comment]
I got same error on a clean project
- Unity 2020.3.12
- Firestore 8.0.0 without other SDKs
- Scripting Runtime: IL2CPP, .Net 4.x
- Android Resolver Settings: "Use jetifer" + "patch gradleTemplate"

**Comment by dconeybe** [comment]
I'm currently investigating, but I wanted to confirm that you have called `Firebase.FirebaseApp.CheckAndFixDependenciesAsync()` prior to using any Firebase APIs.

https://firebase.google.com/docs/reference/unity/class/firebase/firebase-app#class_firebase_1_1_firebase_app_1a330d32bf55497c14bb1393a0a8936868

**Comment by pelevinma** [comment]
Yes, I call this and get `dependencyStatus == Firebase.DependencyStatus.Available`
At least on the previous version of SDK my code is currently running in production

**Comment by dconeybe** [comment]
I've been able to easily reproduce this crash. I didn't even have to do an upgrade. This appears to be a bug in the Unity SDK 8.0.0.

Here is the code that I used to reproduce:
```
using Firebase.Extensions;
using Firebase.Firestore;
using System.Collections.Generic;
using UnityEngine;

public class UnityIssue1080 : MonoBehaviour {

  void Start() {
    Debug.Log("zzyzx CheckAndFixDependenciesAsync() start");
    Firebase.FirebaseApp.CheckAndFixDependenciesAsync().ContinueWithOnMainThread(task => {
      if (task.IsFaulted) {
        Debug.Log("zzyzx CheckAndFixDependenciesAsync() FAILED: " + task.Exception);
        return;
      }

      Debug.Log("zzyzx FirebaseFirestore.DefaultInstance");
      var db = FirebaseFirestore.DefaultInstance;

      DocumentReference docRef = db.Collection("users_v2").Document("1234");
      Dictionary<string, object> userData = new Dictionary<string, object> {
        {"name", "denver"}
      };

      Debug.Log("zzyzx SetAsync()");
      docRef.SetAsync(userData).ContinueWithOnMainThread(task2 => {
        Debug.Log("zzyzx SetAsync() done");
      });
    });
  }

}
```

And the notable error from logcat (full [logcat.txt](https://github.com/firebase/quickstart-unity/files/6741810/logcat.txt)):
```
java.lang.ClassNotFoundException: Didn't find class "com/google/firebase/firestore/internal/cpp/QueryEventListener" on path: DexPathList[[zip file "/data/user/0/com.google.firebase.cpp.firestore.testapp/cache/firestore_resources_lib.jar"],nativeLibraryDirectories=[/system/lib, /system_ext/lib]]
```

I'm actively investigating and will report updates here.

**Comment by dconeybe** [comment]
For some reason [QueryEventListener](https://github.com/firebase/firebase-cpp-sdk/blob/3e86226480704657e6a46b5352be8d453da91e94/firestore/src_java/com/google/firebase/firestore/internal/cpp/QueryEventListener.java) is missing from `firestore_resources_lib.jar`. The rest of the classes, however, are present:

1. CppEventListener                                                                                                                                                                                                                   
1. DocumentEventListener
1. FirestoreTasks
1. JniRunnable
1. LoadBundleProgressListener
1. SilentRejectionSingleThreadExecutor
1. TransactionFunction
1. VoidEventListener

I'm going to try to figure out why `QueryEventListener` is missing.

**Comment by dconeybe** [comment]
Here is a workaround for this bug. I haven't found the root cause yet.  The fix, whatever it is, may not be available for some time.

1. Copy [firestore_cpp_resources.zip](https://github.com/firebase/quickstart-unity/files/6742634/firestore_cpp_resources.zip)
into `Assets/Plugins/Android` of your Unity project.
2. Change the extension from `.zip` to `.jar` (GitHub forbids attaching .jar files, probably for security reasons)
3. Exit and restart the Unity Editor (not sure if this is required, but it fixed a seemingly-unrelated "duplicate import" build error for me)
4. Once the fix is released make sure to delete this JAR file since it will likely conflict with future versions of the Unity SDK.

That `firestore_cpp_resources.jar` file contains the compiled [QueryEventListener](https://github.com/firebase/firebase-cpp-sdk/blob/3e86226480704657e6a46b5352be8d453da91e94/firestore/src_java/com/google/firebase/firestore/internal/cpp/QueryEventListener.java) class that is erroneously absent from the `firestore_resources_lib.jar` that is bundled with the Firebase Unity SDK 8.0.0. By providing in via this JAR file it is able to be found and used.

**Comment by hanesjw** [comment]
Thank you for working on this. I am unfortunately still receiving the same error after trying your solution above.

**Comment by pelevinma** [comment]
stopped crashing with this fix for me

**Comment by dconeybe** [comment]
Thank you both for reporting your experiences with the workaround suggested in my previous comment. I'm continuing investigation but have not found a root cause or fix yet. Unfortunately, due to Canadian and American holidays, and some personal time away from work, I will not be able to work on this issue more until mid-next week. I'll provide an update next week once I have something concrete to report.

**Comment by hanesjw** [comment]
I have no idea why this solution is not working for me. I even made brand new project, imported FirebaseAuth and FirebaseFirestore package. This is what i'm using to try to access the database. It blows up on FirebaseFirestore.DefaultInstance;

`void Start()
    {
        Firebase.FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task => {
            var dependencyStatus = task.Result;
            if (dependencyStatus == Firebase.DependencyStatus.Available)
            {               
                _database = FirebaseFirestore.DefaultInstance;
            }
        });
    }`

I've tried this with a minified build and non-minified build, both with same result. Is there a way to to check if the firebase_cpp_resources.jar file is being included in my apk?

**Comment by Jur-Wind** [comment]
Awesome! Got the same problem and the firestore_cpp_resources.jar fix worked like a charm! Thanks!

**Comment by Toernblom** [comment]
Thanks alot for this! @dconeybe Now i can finally continue developing 😁

**Comment by dconeybe** [comment]
Update: I've found the root cause and a fix is out for review (cl/383270467 for Googlers). There was a typo in our build script that accidentally caused (surprise) the `QueryEventListener` class to be excluded from `firestore_resources_lib.jar`, leading the the crash. The fix was easy and just involved adding `QueryEventListener` back to the manually-maintained list of Java sources from which it was missing. Note that this bug affects the Unity SDK, but not the C++ SDK because the bug was only present in the mirror of the C++ SDK used by the Unity SDK builds. I don't yet have a timeline for when the fix will be released. In the meantime, the workaround suggested in an earlier comment (https://github.com/firebase/quickstart-unity/issues/1080#issuecomment-871572954) should unblock anyone experiencing this crash. Thanks to everyone who reported success with the workaround and for bearing with us as we fix this issue.

**Comment by dconeybe** [comment]
Good news. The fix has been submitted (cl/383683892 for Googlers) and will be included in the next Unity SDK release. Although I cannot commit to a specific date for this release, I would expect it before the end of July. I will add a comment to this issue when that release comes out. In the meantime, please use the workaround from the earlier comment https://github.com/firebase/quickstart-unity/issues/1080#issuecomment-871572954, remembering to "undo" the workaround once the fix is released and incorporated into your project. Thank you all for your patience and apologies again for this critical bug. We will be making changes to our internal processes to avoid errors like this in the future.

**Comment by minh-kage** [comment]
I followed the solution but no luck.
My question is how the build include the firestore_cpp_resources.jar into the android building.

**Comment by Toernblom** [comment]
@minh-kage Try https://convertio.co/zip-jar/ to  convert the .zip to a jar file and place it in the folder mentioned in the previous comments

**Comment by dconeybe** [comment]
Note that a simple rename of the file from `.zip` to `.jar` should do the trick.

**Comment by Toernblom** [comment]
Misleading for some because at default windows 10 it doesn't change the format automatically, see gif:

https://gyazo.com/8574ce69d487a2d51be1e9465cf09d9d

**Comment by dconeybe** [comment]
The fix has just been released in Unity SDK 8.1.0. Remember to undo the workaround (i.e. delete the zip/jar file) when you upgrade.
